### PR TITLE
Added member endpoints for managing newsletter subscriptions

### PR DESCRIPTION
**Pull request by rshbhgrg**
refs TryGhost/Team#1495

With multiple newsletters, members are allowed to manage their newsletter pref via email unsubscribe link with member uuid. Since Portal needs to manage member's newsletter pref via their UUID, we need new endpoints on members that allow fetch/update of newsletter subscriptions via only uuid. The endpoints return only limited data for a member that are needed for the UI.

- adds endpoint to fetch newsletter subscriptions for member via uuid
- adds endpoint to update newsletter subscriptions for member via uuid

### CODE REVIEW

File: core/server/services/members/middleware.js [hunk 1]
```diff
@@ -69,6 +69,57 @@ const getOfferData = async function (req, res) {
     });
 };
 
+const getMemberNewsletters = async function (req, res) {
+    try {
+        const memberUuid = req.query.uuid;
```
Comment (matthanley) [inline core/server/services/members/middleware.js:74]: We should probably guard against this existing before passing it to the members service - I'm not sure what happens if we pass `{uuid: undefined}` to `members.get`

File: core/server/services/members/middleware.js [hunk 2]
```diff
@@ -69,6 +69,57 @@ const getOfferData = async function (req, res) {
...
+            res.writeHead(400);
+            return res.end('Missing member uuid');
+        }
+        const data = _.pick(req.body, 'newsletters');
+        const memberData = await membersService.api.members.get({
+            uuid: memberUuid
+        });
+        if (!memberData) {
+            res.writeHead(400);
+            return res.end('Member not found');
```
Comment (matthanley) [inline core/server/services/members/middleware.js:107]: `get` and `update` should return the same error for non-existence

**Review by matthanley — CHANGES_REQUESTED** [review]

File: test/e2e-frontend/members.test.js [hunk 3]
```diff
@@ -116,6 +116,26 @@ describe('Front-end members behaviour', function () {
...
+
+        it('should error for updating member newsletters with missing uuid', async function () {
+            await request.put('/members/api/member/newsletters')
+                .expect(400);
+        });
+
+        it('should error for updating member newsletters with invalid uuid', async function () {
+            await request.put('/members/api/member/newsletters?uuid=abc')
+                .expect(400);
+        });
```
Comment (ErisDS) [inline test/e2e-frontend/members.test.js:137]: We are missing tests for the success cases. IMO line 59 needs to change from

`await testUtils.initFixtures('members');` 

to

`await testUtils.initFixtures('newsletters', 'members:newsletters');` 

and we should be testing that we get the data we expect?

File: core/server/services/members/middleware.js (hunk 2 above)
Comment (ErisDS) [reply core/server/services/members/middleware.js:107]: non-existence should also be a 404

Comment (rishabhgrg) [reply core/server/services/members/middleware.js]: ✅

File: core/server/services/members/middleware.js (hunk 1 above)
Comment (rishabhgrg) [reply core/server/services/members/middleware.js:74]: ✅

File: core/server/services/members/middleware.js [hunk 4]
```diff
@@ -69,6 +69,57 @@ const getOfferData = async function (req, res) {
...
+        const memberUuid = req.query.uuid;
+
+        const memberData = await membersService.api.members.get({
+            uuid: memberUuid
+        }, {
+            withRelated: ['newsletters']
+        });
+        if (!memberData) {
+            res.writeHead(400);
+            res.end('Email address not found.');
```
Comment (daniellockyer) [inline core/server/services/members/middleware.js:88]: Can we `return` here? If we add more code after this block in the future, the flow will leak through and we'll get the "already sent headers" error

File: core/server/services/members/middleware.js [hunk 5]
```diff
@@ -69,6 +69,62 @@ const getOfferData = async function (req, res) {
...
+            res.writeHead(400);
+            return res.end('Invalid member uuid');
+        }
+        const data = _.pick(req.body, 'newsletters');
+        const memberData = await membersService.api.members.get({
+            uuid: memberUuid
+        });
+        if (!memberData) {
+            res.writeHead(400);
+            res.end('Email address not found.');
```
Comment (daniellockyer) [inline core/server/services/members/middleware.js:112]: This needs a `return;`

File: core/server/services/members/middleware.js (hunk 2 above)
Comment (matthanley) [reply core/server/services/members/middleware.js]: Unresolving for Hannah's 404 comment (in both the get and update)

File: test/e2e-frontend/members.test.js (hunk 3 above)
Comment (rishabhgrg) [reply test/e2e-frontend/members.test.js:139]: @ErisDS Yea totally, updated now, thanks 🙏

File: core/server/services/members/middleware.js (hunk 5 above)
Comment (rishabhgrg) [reply core/server/services/members/middleware.js]: Done, thanks.

File: core/server/services/members/middleware.js (hunk 4 above)
Comment (rishabhgrg) [reply core/server/services/members/middleware.js]: Done, thanks

**Review by matthanley — APPROVED** [review]

File: core/server/services/members/middleware.js (hunk 2 above)
Comment (rishabhgrg) [reply core/server/services/members/middleware.js]: 🙏 Done

**Pull request merged**